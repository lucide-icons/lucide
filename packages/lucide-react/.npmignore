stats
node_modules
tests
scripts
build
src
babel.config.js
jest.config.js
rollup.config.js
