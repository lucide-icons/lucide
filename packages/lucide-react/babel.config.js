const mainConfig = require('../../babel.config');

module.exports = {
  presets: ['react-app'],
  env: mainConfig.env,
};
