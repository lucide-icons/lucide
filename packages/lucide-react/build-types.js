const fs = require("fs")
const path = require("path")
const srcDirectory = path.join(__dirname, "dist")

// Declare type definitions
const typeDefinitions = `
/// <reference types="react" />
import { SVGAttributes } from 'react'

// Create interface extending SVGAttributes
export interface LucideProps extends Partial<React.SVGProps<SVGSVGElement>> {
    color?: string
    size?: string | number
    stroke?: string | number
    strokeWidth?: string | number
}

// Generated icons
`

// Write type definitions to file
fs.writeFileSync(
  path.join(srcDirectory, "index.d.ts"),
  typeDefinitions,
  "utf-8",
)

const iconsFolder = "./build/icons"
fs.readdir(iconsFolder, (err, files) => {

  files.forEach(file => {

    const fileName = file.replace(".js", "")

    // Convert fileName to component name
    const componentName = fileName.split("-")
                                  .map(word => word[0].toUpperCase() + word.substr(1, word.length))
                                  .join("")

    // Declare component type
    const exportTypeString = `export declare const ${componentName}: (props: LucideProps) => JSX.Element;\n`

    // Add component to the types file
    fs.appendFileSync(
      path.join(srcDirectory, "index.d.ts"),
      exportTypeString,
      "utf-8",
    )

  })

  console.log("Generated index.d.ts file with", files.length, "icons")

})
