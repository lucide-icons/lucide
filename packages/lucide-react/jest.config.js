module.exports = {
  verbose: true,
  roots: ['<rootDir>/src/', '<rootDir>/tests/'],
  moduleFileExtensions: ['js'],
  transformIgnorePatterns: [`/node_modules`],
  moduleNameMapper: {
    '^@/(.*)$': '<rootDir>/src/$1',
  },
  transform: {
    '^.+\\.js$': 'babel-jest',
  },
};
