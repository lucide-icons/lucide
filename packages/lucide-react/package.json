{
  "name": "lucide-react",
  "description": "Lucide React package, Lucide is a community-run fork of Feather Icons, open for anyone to contribute icons.",
  "version": "0.15.10",
  "license": "ISC",
  "homepage": "https://lucide.dev",
  "bugs": "https://github.com/lucide-icons/lucide/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/lucide-icons/lucide.git",
    "directory": "packages/lucide-react"
  },
  "author": "Eric Fennis",
  "amdName": "lucide-react",
  "main": "dist/cjs/lucide-react.js",
  "main:umd": "dist/umd/lucide-react.js",
  "module": "dist/esm/lucide-react.js",
  "unpkg": "dist/umd/lucide-react.min.js",
  "typings": "dist/lucide-react.d.ts",
  "scripts": {
    "build": "yarn clean && yarn build:icons && yarn build:es && yarn build:types && yarn build:bundles",
    "clean": "rm -rf dist && rm -rf stats && rm -rf ./src/icons/*.js",
    "build:icons": "yarn --cwd ../../ build:icons --output=../packages/lucide-react/src --templateSrc=../packages/lucide-react/scripts/exportTemplate --renderUniqueKey",
    "build:es": "yarn --cwd ../../ babel packages/lucide-react/src -d packages/lucide-react/dist/esm",
    "build:types": "yarn --cwd ../../ babel-node packages/lucide-react/scripts/buildTypes.js",
    "build:bundles": "yarn --cwd ../../ rollup -c packages/lucide-react/rollup.config.js",
    "test": "jest"
  },
  "dependencies": {
    "prop-types": "^15.7.2",
    "react": "^16.5.1"
  },
  "devDependencies": {
    "babel-preset-react-app": "^10.0.0",
    "jest": "^26.6.3",
    "react-test-renderer": "^16.5.1"
  },
  "peerDependencies": {
    "prop-types": "^15.7.2",
    "react": "^16.5.1"
  }
}
