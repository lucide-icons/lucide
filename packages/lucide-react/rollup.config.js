import plugins from '../../rollup.plugins';
import pkg from './package.json';

const packageName = 'LucideReact';
const outputFileName = 'lucide-react';
const rootDir = 'packages/lucide-react'; // It runs from the root
const outputDir = `${rootDir}/dist`;
const inputs = [`${rootDir}/src/lucide-react.js`];
const bundles = [
  {
    format: 'umd',
    inputs,
    outputDir,
    minify: true,
  },
  {
    format: 'umd',
    inputs,
    outputDir,
  },
  {
    format: 'cjs',
    inputs,
    outputDir,
  },
];

const configs = bundles
  .map(({ inputs, outputDir, format, minify }) =>
    inputs.map(input => ({
      input,
      plugins: plugins(pkg, minify),
      external: ['react', 'prop-types', 'lucide'],
      output: {
        name: packageName,
        file: `${outputDir}/${format}/${outputFileName}${minify ? '.min' : ''}.js`,
        format,
        sourcemap: true,
        globals: {
          react: 'react',
          'prop-types': 'PropTypes',
          lucide: 'lucide',
        },
      },
    })),
  )
  .flat();

export default configs;
