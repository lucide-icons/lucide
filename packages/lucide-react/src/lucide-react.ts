export * from './icons';
export { default as createReactComponent } from './createReactComponent';
