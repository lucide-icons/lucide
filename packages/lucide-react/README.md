# Lucide React

Implementation of the lucide icon library for react applications.

> What is lucide? Read it [here](https://github.com/lucide-icons/lucide#what-is-lucide).

## Installation

```sh
yarn add lucide-react
```

or

```sh
npm install lucide-react
```

## How to use

It's built with ES modules so it's completely tree-shakable.
Each icon can be imported as a react component.

### Example

You can pass additional props to adjust the icon.

```js
import { Camera } from 'lucide-react';

const App = () => {
  return <Camera color="red" size={48} />;
};

export default App;
```

### Props

| name          | type     | default      |
| ------------- | -------- | ------------ |
| `size`        | _Number_ | 24           |
| `color`       | _String_ | currentColor |
| `strokeWidth` | _Number_ | 2            |

### Custom props

You can also pass custom props that will be added in the svg as attributes.

```js
const App = () => {
  return <Camera fill="red" />;
};
```

### Generic icon component

It is possible to create a generic icon component to load icons.

> :warning: The example below is importing all ES modules. This is **not** recommended when you using a bundler since your application build size will grow substantially.

```js
import { icons } from 'lucide-react';

const Icon = ({ name, color, size }) => {
  const LucideIcon = icons[name];

  return <LucideIcon color={color} size={size} />;
};

export default Icon;
```

#### With Dynamic Imports

Lucide react exports a dynamic import map `dynamicIconImports`. Useful for applications that want to show icons dynamically by icon name. For example when using a content management system with where icon names are stored in a database.

When using client side rendering, it will fetch the icon component when it's needed. This will reduce the initial bundle size.

The keys of the dynamic import map are the lucide original icon names.

Example with React suspense:

```tsx
import React, { lazy, Suspense } from 'react';
import { dynamicIconImports, LucideProps } from 'lucide-react';

const fallback = <div style={{ background: '#ddd', width: 24, height: 24 }}/>

interface IconProps extends Omit<LucideProps, 'ref'> {
  name: keyof typeof dynamicIconImports;
}

const Icon = ({ name, ...props }: IconProps) => {
  const LucideIcon = lazy(dynamicIconImports[name]);

  return (
    <Suspense fallback={fallback}>
      <LucideIcon {...props} />
    </Suspense>
  );
}

export default Icon
```

##### NextJS Example

In NextJS, [the dynamic function](https://nextjs.org/docs/pages/building-your-application/optimizing/lazy-loading#nextdynamic) can be used to dynamically load the icon component.

To make dynamic imports work with NextJS, you need to add `lucide-react` to the [`transpilePackages`](https://nextjs.org/docs/app/api-reference/next-config-js/transpilePackages) option in your `next.config.js` like this:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  transpilePackages: ['lucide-react'] // add this
}

module.exports = nextConfig

```

You can then start using it:

```tsx
import dynamic from 'next/dynamic'
import { LucideProps } from 'lucide-react';
import dynamicIconImports from 'lucide-react/dynamicIconImports';

interface IconProps extends LucideProps {
  name: keyof typeof dynamicIconImports;
}

const Icon = ({ name, ...props }: IconProps) => {
  const LucideIcon = dynamic(dynamicIconImports[name])

  return <LucideIcon {...props} />;
};

export default Icon;
```
