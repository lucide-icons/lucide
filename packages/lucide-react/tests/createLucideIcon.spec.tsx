import { describe, it, expect } from 'vitest';
import { createLucideIcon } from '../src/lucide-react';
import { airVent } from './testIconNodes';
import { render } from '@testing-library/react';

describe('Using createLucideIcon', () => {
  it('should create a component from an iconNode', () => {
    const AirVent = createLucideIcon('AirVent', airVent);

    const { container } = render(<AirVent />);

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toBeDefined();
  });
});
