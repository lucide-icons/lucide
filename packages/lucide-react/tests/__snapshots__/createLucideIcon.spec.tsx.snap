// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Using createLucideIcon > should create a component from an iconNode 1`] = `
<svg
  class="lucide lucide-air-vent"
  fill="none"
  height="24"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="2"
  viewBox="0 0 24 24"
  width="24"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M6 12H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v5a2 2 0 0 1-2 2h-2"
  />
  <path
    d="M6 8h12"
  />
  <path
    d="M18.3 17.7a2.5 2.5 0 0 1-3.16 3.83 2.53 2.53 0 0 1-1.14-2V12"
  />
  <path
    d="M6.6 15.6A2 2 0 1 0 10 17v-5"
  />
</svg>
`;
